### PR TITLE
RHDEVDOCS-5245: Performance Tuning for Openshift Pipelines

### PR DESCRIPTION
**Pull request by eromanova97**
• **Aligned team**: Dev Tools
• **OCP version for cherry-picking**:  enterprise-4.12 and later
• **JIRA issue**: [RHDEVDOCS-5245](https://issues.redhat.com/browse/RHDEVDOCS-5245)
• **Preview page**:  [Performance tuning using TektonConfig CR](https://60648--docspreview.netlify.app/openshift-enterprise/latest/cicd/pipelines/installing-pipelines.html#op-performance-tuning-using-tektonconfig-cr_installing-pipelines)
• **SME Review**: Completed by @jkandasa 
• **QE review**: Completed by @ppitonak 
• **Peer-review**: Completed by @mramendi

### CODE REVIEW

**Comment by ekristova** [comment]
Hello @jkandasa @ppitonak PTAL at this PR and provide feedback. Thank you!

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+If you do not specify the fields, the {pipelines-shortname} Operator applies the default values for the {pipelines-shortname} controller.
+
+[NOTE]
+====
+If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `tekton-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed), and recreates `tekton-pipelines-controller` pods.
+====
+
+The following are the modifiable fields:
+
+* `disable-ha`: Enable or disable the HA support. The default value for the pipeline controller is `false`.
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:38]: Why "for the pipeline controller"? And if the value of false means that HA is actually enabled by default, could you maybe explain this here?

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+spec:
+  performance:
+    disable-ha: false
+    buckets: 1
+    threads-per-controller: 2
+    kube-api-qps: 5.0
+    kube-api-burst: 10
+----
+
+The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `tekton-pipelines-controller` deployment under the `tekton-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `tekton-pipelines` namespace.
```
Comment (jkandasa) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:27]: namespace in product will be `openshift-pipelines`
`tekton-pipelines` should be as `openshift-pipelines`

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+If you do not specify the fields, the {pipelines-shortname} Operator applies the default values for the {pipelines-shortname} controller.
```
Comment (jkandasa) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:29]: if you do not specify the values, operator will not update those fields. default values from pipelines-controller will be used.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
++
+The given number of replicas uses the specified number of buckets. The owner of a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` and the default value in the pipeline controller is `1`.
++
+[NOTE]
+====
+The {pipelines-shortname} Operator does not handle the pipeline controller deployment of `replicas`. For example, to change the number of replicas to 3, run:
+
+[source,terminal]
+----
+$ oc --namespace tekton-pipelines scale deployment tekton-pipelines-controller --replicas=3
```
Comment (jkandasa) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:50]: namepsace => `openshift-pipelines`

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 5]
```diff
@@ -0,0 +1,63 @@
+// This module is included in the following assembly:
+//
+// *openshift_pipelines/customizing-configurations-in-the-tektonconfig-cr.adoc
+
+:_content-type: CONCEPT
+[id="op-performance-tuning-using-tektonconfig-cr_{context}"]
+= Performance tuning using TektonConfig CR
+
+You can modify the fields under the `.spec.performance` parameter in the `TektonConfig` custom resource (CR) to change high availability (HA) support and performance configurations for the {pipelines-shortname} controller.
```
Comment (jkandasa) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:9]: `.spec.pipeline.performance` parameter in the `TektonConfig`

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+kind: TektonConfig
+metadata:
+  name: config
+spec:
+  performance:
+    disable-ha: false
+    buckets: 1
+    threads-per-controller: 2
+    kube-api-qps: 5.0
+    kube-api-burst: 10
```
Comment (jkandasa) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:24]:
```yaml
spec:
  pipeline:
    performance:
      disable-ha: false
      buckets: 1
      threads-per-controller: 2
      kube-api-qps: 5.0
      kube-api-burst: 10
```

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `tekton-pipelines-controller` deployment under the `tekton-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `tekton-pipelines` namespace.
+
+If you do not specify the fields, the {pipelines-shortname} Operator applies the default values for the {pipelines-shortname} controller.
+
+[NOTE]
+====
+If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `tekton-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed), and recreates `tekton-pipelines-controller` pods.
+====
+
+The following are the modifiable fields:
```
[the remark this answers is not in the document]
Comment (ekristova) [reply modules/op-performance-tuning-using-tektonconfig-cr.adoc:37]: I put it in the table, but for the table to not look squeezed, I had to move the notes outside of the table.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc (hunk 1 above)
Comment (ekristova) [reply modules/op-performance-tuning-using-tektonconfig-cr.adoc:39]: @mramendi I wrote "for the pipeline controller" to specify that the fields are optional. The default values therefore indicate what is the default configuration for the pipeline controller. Do you think the specification here is unnecessary?

**Review by ppitonak — APPROVED** [review]

**Review by jkandasa — APPROVED** [review]
LGTM

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+[options="header"]
+|===
+
+| Name | Description | Default value for the pipeline controller
+
+| `disable-ha` | Enable or disable the HA support. By default, the HA support is enabled. | `false`
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:44]: Suggestion: expand "HA" to "High Availability (HA)" at first mention.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+| `buckets` | Indicates the number of buckets used to partition key space of each reconciler.
+
+The given number of replicas uses the specified number of buckets. The owner of a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:48]: Could we still expand "owner" to "the instance that owns"? Usually "owner" is a user

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (mramendi) [reply modules/op-performance-tuning-using-tektonconfig-cr.adoc:37]: All good on this one!

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      kube-api-burst: 10
+----
+
+The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `openshift-pipelines-controller` deployment under the `openshift-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `openshift-pipelines` namespace.
+
+If you do not specify the values, the {pipelines-shortname} Operator does not update those fields and applies the default values for the {pipelines-shortname} controller.
+
+[NOTE]
+====
+If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `openshift-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed), and recreates `openshift-pipelines-controller` pods.
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:34]: Suggest removing the comma before "and" and adding a dash "re-creates".

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+====
+
+The following are the modifiable fields:
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:37]: Suggest making this a table title instead (a .something right before the table, no empty lines, turns into a table title) and rewording:

.Configuration fields for tuning {pipelines-shortname} performance

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+[options="header"]
+|===
+
+| Name | Description | Default value for the pipeline controller
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:42]: {pipelines-shortname} controller

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+| `disable-ha` | Enable or disable the high availability (HA) support. By default, the HA support is enabled. | `false`
+
+| `buckets` | Indicates the number of buckets used to partition key space of each reconciler.
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:46]: Remove "Indicates" - just "The number...". ("Indicates" is just outright wrong, this is a configuration field not a monitoring field). I would prefer to remove the passive voice if possible.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+The given number of replicas uses these buckets. The instance that owns a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:48]: "Each of the replicas" or "All of the replicas, combined" - depending on which is true. "The given number" is still somewhat unclear.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+| `threads-per-controller` | Specifies the number of threads (workers) to use when the workqueue of the pipeline controller is proccessed. | `2`
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:50]: typo - "processed".

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+[source,terminal]
+----
+$ oc --namespace openshift-pipelines scale deployment openshift-pipelines-controller --replicas=3
+----
+====
+
+[IMPORTANT]
+====
+The `kube-api-qps` and `kube-api-burst` fields are multiplied by 2 in the pipeline controller. For example, if the `kube-api-qps` and `kube-api-burst` values are `10`, the actual QPS and burst value becomes `20`.
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:70]: "become" not "becomes" as "values" is plural

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc (hunk 10 above)
Comment (ekristova) [reply modules/op-performance-tuning-using-tektonconfig-cr.adoc]: I incorporated the rest of the suggestions.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+| Name | Description | Default value for the {pipelines-shortname} controller
+
+| `disable-ha` | Enable or disable the high availability (HA) support. By default, the HA support is enabled. | `false`
+
+| `buckets` | The number of buckets used to partition key space of each reconciler.
+
+Each of the replicas uses these buckets. The instance that owns a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`
+
+| `threads-per-controller` | Specifies the number of threads (workers) to use when the workqueue of the {pipelines-shortname} controller is processed. | `2`
```
Comment (mramendi) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:49]: Remove "specifies" here and "indicates" in the next one, please

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc (hunk 10 above)
Comment (mramendi) [reply modules/op-performance-tuning-using-tektonconfig-cr.adoc]: I have sent the IBM Style Guide reference in Slack as the document is not available to the general public.

**Review by mramendi — APPROVED** [review]

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+.Modifiable fields for tuning {pipelines-shortname} performance
+[options="header"]
+|===
+
+| Name | Description | Default value for the {pipelines-shortname} controller
+
+| `disable-ha` | Enable or disable the high availability (HA) support. By default, the HA support is enabled. | `false`
+
+| `buckets` | The number of buckets used to partition key space of each reconciler.
```
Comment (bburt-rh) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:45]:
```suggestion
| `buckets` | The number of buckets used to partition the key space for each reconciler.
```
- add definite article
- 'for' is more idiomatic here

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+Each of the replicas uses these buckets. The instance that owns a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`
```
Comment (bburt-rh) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:47]:
```suggestion
Each of the replicas uses these buckets. The instance that owns a bucket reconciles the keys partitioned into that bucket. The maximum value is `10` | `1`
```
- fix awkward noun-based phrasing `reconciling of`

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+| `threads-per-controller` | The number of threads (workers) to use when the workqueue of the {pipelines-shortname} controller is processed. | `2`
```
Comment (bburt-rh) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:49]: Should `workqueue` be `work queue`? If `workqueue` isn't the name of a technical feature, it should be two words.

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+| `kube-api-qps` | The maximum queries per second (QPS) to the cluster master from the REST client. | `5.0`
+
+| `kube-api-burst` | The maximum burst for throttle. | `10`
```
Comment (bburt-rh) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:53]:
```suggestion
| `kube-api-burst` | The maximum burst for a throttle. | `10`
```

File: modules/op-performance-tuning-using-tektonconfig-cr.adoc [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+//
+// *openshift_pipelines/customizing-configurations-in-the-tektonconfig-cr.adoc
+
+:_content-type: CONCEPT
+[id="op-performance-tuning-using-tektonconfig-cr_{context}"]
+= Performance tuning using TektonConfig CR
+
+You can modify the fields under the `.spec.pipeline.performance` parameter in the `TektonConfig` custom resource (CR) to change high availability (HA) support and performance configuration for the {pipelines-shortname} controller.
+
+.Example: `TektonConfig` performance fields
```
Comment (bburt-rh) [inline modules/op-performance-tuning-using-tektonconfig-cr.adoc:11]:
```suggestion
.Example: TektonConfig performance fields
```
Headings can't have backticks or other markup. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings

**Comment by shipsing** [comment]
Cherry-pick the content to enterprise 4.12, 4.13, and 4.14

**Pull request merged**